In [143]:
from sympy import *
import numpy as np

D = 3

## Definiendo simbolos
t, r, theta, phi = symbols('t, r, \\theta, \\varphi')
dt, dr, dtheta, dphi = symbols('dt dr dtheta dphi')     
l0, l1, l2, l3 = symbols("l_0 l_1 l_2 l_3")
l, lamb, p, q, n, L = symbols("l lamb p q n L")

deldel = Function("DELDEL")
delf = Function("DELF")
box = Function("\\Box")

def getFuncName(i):
    if i==0:
        return alfa.name
    if i==1:
        return F.name
    if i==2:
        return G.name
    
    
if D == 1:
    ndmat = np.array([[[symbols("a") for _x in range(4)] for _w in range(4)] for _i in range(3)])
    alfa = Function("a")(t, r)
    F = Function("F")(t, r)
    G = Function("G")(t, r)
elif D == 2:
    ndmat = np.array([[[[symbols("a") for _x in range(4)] for _y in range(4)] for _w in range(4)] for _i in range(3)])
    alfa = Function("a")(t, theta, r)
    F = Function("F")(t, theta, r)
    G = Function("G")(t, theta, r)
elif D == 3:
    ndmat = np.array([[[[[symbols("a") for _x in range(4)] for _y in range(4)] for _z in range(4)] for _w in range(4)] for _i in range(3)])
    alfa = Function("a")(t, phi, theta, r)
    F = Function("F")(t, phi, theta, r)
    G = Function("G")(t, phi, theta, r)


symbolsString = ""
if D == 1:
    for _f in range(3):
        for _t in range(4):
            for _r in range(4):
                ndmat[_f][_t][_r] = symbols(getFuncName(_f)+str(_t)+str(_r))
                symbolsString += str(ndmat[_f][_t][_r])+" "
elif D == 2:
    for _f in range(3):
        for _t in range(4):
            for _theta in range(4):
                for _r in range(4):
                    ndmat[_f][_t][_theta][_r] = symbols(getFuncName(_f)+str(_t)+str(_theta)+str(_r))
                    symbolsString += str(ndmat[_f][_t][_theta][_r])+" "
elif D == 3:
    for _f in range(3):
        for _t in range(4):
            for _phi in range(4):
                for _theta in range(4):
                    for _r in range(4):
                        ndmat[_f][_t][_phi][_theta][_r] = symbols(getFuncName(_f)+str(_t)+str(_phi)+str(_theta)+str(_r))
                        symbolsString += str(ndmat[_f][_t][_phi][_theta][_r])+" "


In [144]:
symbolsString

'a000 a001 a002 a003 a010 a011 a012 a013 a020 a021 a022 a023 a030 a031 a032 a033 a100 a101 a102 a103 a110 a111 a112 a113 a120 a121 a122 a123 a130 a131 a132 a133 a200 a201 a202 a203 a210 a211 a212 a213 a220 a221 a222 a223 a230 a231 a232 a233 a300 a301 a302 a303 a310 a311 a312 a313 a320 a321 a322 a323 a330 a331 a332 a333 F000 F001 F002 F003 F010 F011 F012 F013 F020 F021 F022 F023 F030 F031 F032 F033 F100 F101 F102 F103 F110 F111 F112 F113 F120 F121 F122 F123 F130 F131 F132 F133 F200 F201 F202 F203 F210 F211 F212 F213 F220 F221 F222 F223 F230 F231 F232 F233 F300 F301 F302 F303 F310 F311 F312 F313 F320 F321 F322 F323 F330 F331 F332 F333 G000 G001 G002 G003 G010 G011 G012 G013 G020 G021 G022 G023 G030 G031 G032 G033 G100 G101 G102 G103 G110 G111 G112 G113 G120 G121 G122 G123 G130 G131 G132 G133 G200 G201 G202 G203 G210 G211 G212 G213 G220 G221 G222 G223 G230 G231 G232 G233 G300 G301 G302 G303 G310 G311 G312 G313 G320 G321 G322 G323 G330 G331 G332 G333 '

In [145]:
import sys
sys.version

'3.9.1 (default, Feb  6 2021, 06:49:13) \n[GCC 10.2.0]'

In [146]:
def DELDEL(f, g):
    #canon = -f.diff(t)*g.diff(t) + 1/l1**2*f.diff(r)*g.diff(r) + 1/l1**2*f.diff(theta)*g.diff(theta) + 1/l2**2*f.diff(phi)*g.diff(phi)
    canon = -f.diff(t)*g.diff(t) + f.diff(r)*g.diff(r) + f.diff(theta)*g.diff(theta) + f.diff(phi)*g.diff(phi)
    return canon

def BOX(f):
    canon = -f.diff(t, 2) + (f.diff(r, 2) + f.diff(theta, 2) + f.diff(phi, 2))
    #canon = -f.diff(t, 2) + 1/l1**2*f.diff(r, 2) + 1/l1**2*f.diff(theta, 2) + 1/l2 **2*f.diff(phi, 2)
    return canon

def DELf(f, g):
    #canon = DELDEL(f, g) + f*BOX(g)

    canon = -(f*g.diff(t)).diff(t) + (f*g.diff(r)).diff(r)+ (f*g.diff(theta)).diff(theta) + (f*g.diff(phi)).diff(phi)
    #canon = -(f*g.diff(t)).diff(t) + 1/l1**2*(f*g.diff(r)).diff(r)+ 1/l1**2*(f*g.diff(theta)).diff(theta) + 1/l2**2*(f*g.diff(phi)).diff(phi)
    return canon


with evaluate(False):
    motion_alfa = box(alfa) + 2*(1 + 1/2*t)*(theta*(1-theta) + r*(1-r))
    
    motion_F = box(F)
    
    motion_G = box(G)
    
display(motion_alfa)
display(motion_F)
display(motion_G)

(2*(0.5*t + 1))*(\theta*(1 - \theta) + r*(1 - r)) + \Box(a(t, \theta, r))

\Box(F(t, \theta, r))

\Box(G(t, \theta, r))

In [147]:
def expandToDerivatives(func):
    p, q, s, b = map(Wild, 'pqsb')
    res = func.replace(deldel(p, q)*s, lambda p, q, s : DELDEL(p, q)*s)
    res = res.replace(delf(p, q)*s, lambda p, q, s : DELf(p, q)*s)
    return res.replace(box(p)*s, lambda p, s : BOX(p)*s)

In [148]:
res3a = (expandToDerivatives(motion_alfa))
display("Alfa done.")
res3f = (expandToDerivatives(motion_F))
display("F done.")
res3g = (expandToDerivatives(motion_G))
display("G done.")

'Alfa done.'

'F done.'

'G done.'

In [149]:
display(res3a)
display(res3f)
display(res3g)

(2*(0.5*t + 1))*(\theta*(1 - \theta) + r*(1 - r)) + Derivative(a(t, \theta, r), (\theta, 2)) + Derivative(a(t, \theta, r), (r, 2)) - Derivative(a(t, \theta, r), (t, 2))

Derivative(F(t, \theta, r), (\theta, 2)) + Derivative(F(t, \theta, r), (r, 2)) - Derivative(F(t, \theta, r), (t, 2))

Derivative(G(t, \theta, r), (\theta, 2)) + Derivative(G(t, \theta, r), (r, 2)) - Derivative(G(t, \theta, r), (t, 2))

In [154]:
def getDelta(i):
    if i == t:
        return dt
    if r == i:
        return dr
    if theta == i:
        return dtheta
    if phi == i:
        return dphi

def getIndex(i):
    if t == i:
        return 0
    if phi == i:
        return 1
    if theta == i:
        return 2
    if r == i:
        return 3

def getFuncIndex(f):
    if f==F:
        return 1
    if f==G:
        return 2
    if f==alfa:
        return 0
    
def isThis(s, ix):
    if ix != None and s in ix.keys():
        return ix[s]
    else: 
        return 0
#http://www.iitg.ac.in/gtm/files/chap2.pdf
#https://www.uio.no/studier/emner/matnat/math/MAT-INF1100/h07/undervisningsmateriale/kap7.pdf

# 3: +1
# 2:  0
# 1: -1
# 0: -2

def getDiscreteSymbol(ndmat, f, ix=None):
    if D==3:
        canon = ndmat[getFuncIndex(f)][2+isThis(t, ix)][2+isThis(phi, ix)][2+isThis(theta, ix)][2+isThis(r, ix)]
    elif D==2:
        canon = ndmat[getFuncIndex(f)][2+isThis(t, ix)][2+isThis(theta, ix)][2+isThis(r, ix)]
    elif D==1:
        canon = ndmat[getFuncIndex(f)][2+isThis(t, ix)][2+isThis(r, ix)]
    return canon
    
def FOD(f, i, wrt):
    if i==t:
        canon = (getDiscreteSymbol(ndmat, f, ix={i:0}) - getDiscreteSymbol(ndmat, f, ix={i:-1}))/(getDelta(i))
    else:
        canon = (getDiscreteSymbol(ndmat, f, ix={i:0}) - getDiscreteSymbol(ndmat, f, ix={i:-1}))/(getDelta(i))
    return canon

def SOD(f, i, j, wrt):
    if i==j:
        if (f==wrt and i==t) or i != t:
            canon = (getDiscreteSymbol(ndmat, f, ix={i:1}) - 2*getDiscreteSymbol(ndmat, f, ix={i:0}) + getDiscreteSymbol(ndmat, f, ix={i:-1}))/(getDelta(i)**2)
        else:
            canon = (getDiscreteSymbol(ndmat, f, ix={i:0}) - 2*getDiscreteSymbol(ndmat, f, ix={i:-1}) + getDiscreteSymbol(ndmat, f, ix={i:-2}))/(getDelta(i)**2)


    else:
        if i != t and t != j:
            canon = (getDiscreteSymbol(ndmat, f, ix={i:1, j:1}) + getDiscreteSymbol(ndmat, f, ix={i:-1, j:-1}) - getDiscreteSymbol(ndmat, f, ix={i:1, j:-1}) - getDiscreteSymbol(ndmat, f, ix={i:-1, j:1}) )/(4*getDelta(i)*getDelta(j))
        #backward difference
        elif (j==t):
            canon = (1/getDelta(i))*(((getDiscreteSymbol(ndmat, f, ix={i:1, j:0})-getDiscreteSymbol(ndmat, f, ix={i:0, j:0}))/(getDelta(j))) - ((getDiscreteSymbol(ndmat, f, ix={i:1, j:-1}) - getDiscreteSymbol(ndmat, f, ix={i:0, j:-1}))/(getDelta(j))))
            
            #canon = (1/(2*getDelta(i)*getDelta(j)))*(getDiscreteSymbol(ndmat, f, ix={i:0, j:1})-getDiscreteSymbol(ndmat, f, ix={i:-1, j:1})-getDiscreteSymbol(ndmat, f, ix={i:0, j:-1})-getDiscreteSymbol(ndmat, f, ix={i:-1, j:-1}))
        elif (i == t):
            #canon = (1/(2*getDelta(i)*getDelta(j)))*(getDiscreteSymbol(ndmat, f, ix={i:1, j:0})-getDiscreteSymbol(ndmat, f, ix={i:1, j:-1})-getDiscreteSymbol(ndmat, f, ix={i:-1, j:0})-getDiscreteSymbol(ndmat, f, ix={i:-1, j:-1}))
            canon = (1/getDelta(i))*(((getDiscreteSymbol(ndmat, f, ix={i:0, j:1})-getDiscreteSymbol(ndmat, f, ix={i:0, j:0}))/(getDelta(j))) - ((getDiscreteSymbol(ndmat, f, ix={i:-1, j:1}) - getDiscreteSymbol(ndmat, f, ix={i:-1, j:0}))/(getDelta(j))))
            
        else:
            print("what")
    return canon

def discreti(arg, wrt):
    if len(arg.variables) == 1:
        return FOD(arg.expr, arg.variables[0], wrt)
    elif len(arg.variables) == 2:
        return SOD(arg.expr, arg.variables[0], arg.variables[1], wrt)
    else:
        return None

In [155]:
def discretizeVariables(func):
    res = func.replace(r, r*dr)
    res = res.replace(theta, theta*dtheta)
    res = res.replace(phi, phi*dphi)
    return res.replace(t, l*dt)

def discretizeDerivatives(func, wrt):
    res = func.replace(lambda arg: arg.is_Derivative, lambda arg: discreti(arg, wrt))
    res = res.replace(F, getDiscreteSymbol(ndmat, F))
    res = res.replace(G, getDiscreteSymbol(ndmat, G))
    return res.replace(alfa, getDiscreteSymbol(ndmat, alfa))

In [156]:
res7a = discretizeDerivatives(res3a, alfa)
res7a = discretizeVariables(res7a)

res7f = discretizeDerivatives(res3f, F)
res7f = discretizeVariables(res7f)

res7g = discretizeDerivatives(res3g, G)
res7g = discretizeVariables(res7g)


In [157]:
display(res7a)
display(res7f)
display(res7g)

2*(0.5*dt*l + 1)*(\theta*dtheta*(-\theta*dtheta + 1) + dr*r*(-dr*r + 1)) + (a212 - 2*a222 + a232)/dtheta**2 - (a122 - 2*a222 + a322)/dt**2 + (a221 - 2*a222 + a223)/dr**2

(F212 - 2*F222 + F232)/dtheta**2 - (F122 - 2*F222 + F322)/dt**2 + (F221 - 2*F222 + F223)/dr**2

(G212 - 2*G222 + G232)/dtheta**2 - (G122 - 2*G222 + G322)/dt**2 + (G221 - 2*G222 + G223)/dr**2

In [158]:
expra = ccode(res7a).replace("\\", "")
exprf = ccode(res7f).replace("\\", "")
exprg = ccode(res7g).replace("\\", "")

In [159]:
expra

'2*(0.5*dt*l + 1)*(theta*dtheta*(-theta*dtheta + 1) + dr*r*(-dr*r + 1)) + (a212 - 2*a222 + a232)/pow(dtheta, 2) - (a122 - 2*a222 + a322)/pow(dt, 2) + (a221 - 2*a222 + a223)/pow(dr, 2)'

In [160]:
def getSymbolsUsed(expr):
    syms = ""
    for _y in expr.free_symbols:
        syms += str(_y)+" "
    return syms[:-1]

In [161]:
symma = getSymbolsUsed(res7a)
print("Symbols for a: \n [", symma, "]")

symmf = getSymbolsUsed(res7f)
print("Symbols for F: \n [", symmf, "]")

symmg = getSymbolsUsed(res7g)
print("Symbols for G: \n [", symmg, "]")


Symbols for a: 
 [ dr a212 a122 l a322 \theta a221 a223 r a232 dtheta dt a222 ]
Symbols for F: 
 [ dr dt F223 F122 F221 F222 F232 dtheta F322 F212 ]
Symbols for G: 
 [ dr G221 G212 G122 G223 G222 dtheta G232 dt G322 ]


In [162]:
#import pyximport
#pyximport.install()
from sage.all import *
import sage as sg

var(symma.replace("\\", ""))
var(symmf.replace("\\", ""))
var(symmg.replace("\\", ""))
var(symbolsString.replace("\\", ""))
var("E4 E5 E6 r theta global_phi")

pi_4 = E4*sin(dr*r)*sin(dtheta*theta)*sin(dphi*global_phi)
pi_5 = E5*sin(dr*r)*sin(dtheta*theta)*sin(dphi*global_phi)
pi_6 = E6*sin(dr*r)*sin(dtheta*theta)*sin(dphi*global_phi)

#dAdt = "0 + pi_4"
dAdt = "0.5*(dr*r*(1 - dr*r)*dtheta*theta*(1 - dtheta*theta))"



dFdt = "0 + pi_5"
dGdt = "p/L + pi_6"

def solveWithSage(expr, symbols, x):
    print(expr)
    wata = eval(expr)
    resa = sg.symbolic.relation.solve(wata, x)
    return resa[0]

In [163]:
def getDerivative(eq, var):
    if type(eq) == int:
        return 0
    else:
        try:
            return eq.derivative(var)
        except:
            return 0


def replaceOtherFunctionsNegative(expr, func, func1, func2, ofunc1dt, ofunc2dt):
    auxofun1 = eval(func1+"2222") - 2.0*dt*ofunc1dt 
    auxofun2 = eval(func2+"2222") - 2.0*dt*ofunc2dt 
    
    firsta = expr.subs({eval(func1+"0222") : auxofun1})
    firsta = firsta.subs({eval(func2+"0222") : auxofun2})
    firstaa = sg.symbolic.relation.solve(firsta, eval(func+"3222"))
    try:
        return firstaa[0]#.collect_common_factors().full_simplify().collect_common_factors()
    except:
        return firstaa


In [165]:
from sage.misc.latex import Latex
ra=[]
if D==3:
    ra = solveWithSage(expra, symma, a3222)
if D==2:
    ra = solveWithSage(expra, symma, a322)
if D==1:
    ra = solveWithSage(expra, symma, a32)
lates = Latex()
lates(ra)

2*(0.5*dt*l + 1)*(theta*dtheta*(-theta*dtheta + 1) + dr*r*(-dr*r + 1)) + (a212 - 2*a222 + a232)/pow(dtheta, 2) - (a122 - 2*a222 + a322)/pow(dt, 2) + (a221 - 2*a222 + a223)/pow(dr, 2)


a_{322} = \frac{{\left(a_{212} - 2 \, a_{222} + a_{232}\right)} \mathit{dr}^{2} \mathit{dt}^{2} - {\left({\left(a_{122} - 2 \, a_{222}\right)} \mathit{dr}^{2} - {\left(a_{221} - 2 \, a_{222} + a_{223}\right)} \mathit{dt}^{2}\right)} \mathit{dtheta}^{2} - {\left(\mathit{dr}^{4} \mathit{dt}^{3} \mathit{dtheta}^{2} l + 2 \, \mathit{dr}^{4} \mathit{dt}^{2} \mathit{dtheta}^{2}\right)} r^{2} - {\left(\mathit{dr}^{2} \mathit{dt}^{3} \mathit{dtheta}^{4} l + 2 \, \mathit{dr}^{2} \mathit{dt}^{2} \mathit{dtheta}^{4}\right)} \theta^{2} + {\left(\mathit{dr}^{3} \mathit{dt}^{3} \mathit{dtheta}^{2} l + 2 \, \mathit{dr}^{3} \mathit{dt}^{2} \mathit{dtheta}^{2}\right)} r + {\left(\mathit{dr}^{2} \mathit{dt}^{3} \mathit{dtheta}^{3} l + 2 \, \mathit{dr}^{2} \mathit{dt}^{2} \mathit{dtheta}^{3}\right)} \theta}{\mathit{dr}^{2} \mathit{dtheta}^{2}}

$a_{322} = \frac{{\left(a_{212} - 2 \, a_{222} + a_{232}\right)} \mathit{dr}^{2} \mathit{dt}^{2} - {\left({\left(a_{122} - 2 \, a_{222}\right)} \mathit{dr}^{2} - {\left(a_{221} - 2 \, a_{222} + a_{223}\right)} \mathit{dt}^{2}\right)} \mathit{dtheta}^{2} - {\left(\mathit{dr}^{4} \mathit{dt}^{3} \mathit{dtheta}^{2} l + 2 \, \mathit{dr}^{4} \mathit{dt}^{2} \mathit{dtheta}^{2}\right)} r^{2} - {\left(\mathit{dr}^{2} \mathit{dt}^{3} \mathit{dtheta}^{4} l + 2 \, \mathit{dr}^{2} \mathit{dt}^{2} \mathit{dtheta}^{4}\right)} \theta^{2} + {\left(\mathit{dr}^{3} \mathit{dt}^{3} \mathit{dtheta}^{2} l + 2 \, \mathit{dr}^{3} \mathit{dt}^{2} \mathit{dtheta}^{2}\right)} r + {\left(\mathit{dr}^{2} \mathit{dt}^{3} \mathit{dtheta}^{3} l + 2 \, \mathit{dr}^{2} \mathit{dt}^{2} \mathit{dtheta}^{3}\right)} \theta}{\mathit{dr}^{2} \mathit{dtheta}^{2}}$

In [166]:
def nops(expr):
    count = 0
    string = ccode(expr._sympy_())
    count += string.count("*")
    count += string.count("^")
    count += string.count("-")
    count += string.count("+")
    count += string.count("/")
    return count

print(nops(ra))
#print(nops(rf))
#print(nops(rg))

53


In [167]:
newra = replaceOtherFunctionsNegative(ra, "a", "F", "G", eval(dFdt), eval(dGdt))
print(newra)

[

]


In [169]:
rf = []
if D==3:
    rf = solveWithSage(exprf, symmf, F3222)
elif D==2:
    rf = solveWithSage(exprf, symmf, F322)
elif D==1:
    rf = solveWithSage(exprf, symmf, F32)
print(rf)

(F212 - 2*F222 + F232)/pow(dtheta, 2) - (F122 - 2*F222 + F322)/pow(dt, 2) + (F221 - 2*F222 + F223)/pow(dr, 2)
F322 == ((F212 - 2*F222 + F232)*dr^2*dt^2 - ((F122 - 2*F222)*dr^2 - (F221 - 2*F222 + F223)*dt^2)*dtheta^2)/(dr^2*dtheta^2)


In [170]:
newrf = replaceOtherFunctionsNegative(rf, "F", "a", "G", eval(dAdt), eval(dGdt))
print(newrf)

[

]


In [171]:
rg = []
if D==3:
    rg = solveWithSage(exprg, symmg, G3222)
elif D==2:
    rg = solveWithSage(exprg, symmg, G322)
elif D==1:
    rg = solveWithSage(exprg, symmg, G32)
print(rg)

(G212 - 2*G222 + G232)/pow(dtheta, 2) - (G122 - 2*G222 + G322)/pow(dt, 2) + (G221 - 2*G222 + G223)/pow(dr, 2)
G322 == ((G212 - 2*G222 + G232)*dr^2*dt^2 - ((G122 - 2*G222)*dr^2 - (G221 - 2*G222 + G223)*dt^2)*dtheta^2)/(dr^2*dtheta^2)


In [172]:
newrg = replaceOtherFunctionsNegative(rg, "G", "a", "F", eval(dAdt), eval(dFdt))
print(newrg)

[

]


In [260]:
import re
def i2i(arg, index):
    if index == "0":
        return "("+arg+"-2)"
    if index == "1":
        return "("+arg+"-1)"
    if index == "2":
        return "("+arg+")"
    if index == "3":
        return "("+arg+"+1)"
def i2iTime(arg, index):
    if index == "0":
        return "(tm2)"
    if index == "1":
        return "(tm1)"
    if index == "2":
        return "(t)"
    if index == "3":
        return "(tp1)"    
def getCArray(ids):
    res = ""
    args = ["t", "threadIdx.z", "threadIdx.y", "threadIdx.x"]
    sizes = ["L", "O", "N", "M"]  
    while len(ids)!=4:
        ids = ids[0]+'2'+ids[1:]
    for i, c in enumerate(ids):
        if args[i] == "t":
            res += i2iTime(args[i], c)+","
        else:
            res += i2i(args[i], c)+","
        #for j,s in enumerate(sizes[i+1:]):
            #res+="*"+s
        #res+="+"
    return res[:-1]


def getExpressionInC(expr):
    
    traduc = ccode(expr)

    if D==3:
        indexes = [m.start() for m in re.finditer(r'[a-zA-Z]\d\d\d\d', traduc)]
        for i in reversed(indexes):
            traduc = traduc[:i+1]+"[I("+ getCArray(traduc[i+1:i+5]) +")]"+ traduc[i+5:]
    elif D==2:
        indexes = [m.start() for m in re.finditer(r'[a-zA-Z]\d\d\d', traduc)]
        for i in reversed(indexes):
            traduc = traduc[:i+1]+"[I("+ getCArray(traduc[i+1:i+4]) +")]"+ traduc[i+4:]
    elif D==1:
        indexes = [m.start() for m in re.finditer(r'[a-zA-Z]\d\d', traduc)]
        for i in reversed(indexes):
            traduc = traduc[:i+1]+"[I("+ getCArray(traduc[i+1:i+3]) +")]"+ traduc[i+3:]
    traduc = traduc.replace("a[I(", "shmem[CI(0,")
    traduc = traduc.replace("F[I(", "shmem[CI(1,")
    traduc = traduc.replace("G[I(", "shmem[CI(2,")
    
    traduc = traduc.replace("threadIdx.x+1", "tidxp")
    traduc = traduc.replace("threadIdx.x-1", "tidxm")
    traduc = traduc.replace("threadIdx.y+1", "tidyp")
    traduc = traduc.replace("threadIdx.y-1", "tidym")
    traduc = traduc.replace("threadIdx.z+1", "tidzp")
    traduc = traduc.replace("threadIdx.z-1", "tidzm")    


    return traduc

In [261]:
expr_a = getExpressionInC(ra._sympy_())
expr_a

'shmem[CI(0,(tp1),(threadIdx.z),(threadIdx.y),(threadIdx.x))] == (pow(dr, 2)*pow(dt, 2)*(shmem[CI(0,(t),(threadIdx.z),(tidym),(threadIdx.x))] - 2*shmem[CI(0,(t),(threadIdx.z),(threadIdx.y),(threadIdx.x))] + shmem[CI(0,(t),(threadIdx.z),(tidyp),(threadIdx.x))]) - pow(dtheta, 2)*(pow(dr, 2)*(shmem[CI(0,(tm1),(threadIdx.z),(threadIdx.y),(threadIdx.x))] - 2*shmem[CI(0,(t),(threadIdx.z),(threadIdx.y),(threadIdx.x))]) - pow(dt, 2)*(shmem[CI(0,(t),(threadIdx.z),(threadIdx.y),(tidxm))] - 2*shmem[CI(0,(t),(threadIdx.z),(threadIdx.y),(threadIdx.x))] + shmem[CI(0,(t),(threadIdx.z),(threadIdx.y),(tidxp))])) - pow(r, 2)*(pow(dr, 4)*pow(dt, 3)*pow(dtheta, 2)*l + 2*pow(dr, 4)*pow(dt, 2)*pow(dtheta, 2)) + r*(pow(dr, 3)*pow(dt, 3)*pow(dtheta, 2)*l + 2*pow(dr, 3)*pow(dt, 2)*pow(dtheta, 2)) - pow(theta, 2)*(pow(dr, 2)*pow(dt, 3)*pow(dtheta, 4)*l + 2*pow(dr, 2)*pow(dt, 2)*pow(dtheta, 4)) + theta*(pow(dr, 2)*pow(dt, 3)*pow(dtheta, 3)*l + 2*pow(dr, 2)*pow(dt, 2)*pow(dtheta, 3)))/(pow(dr, 2)*pow(dtheta, 2))'

In [262]:
expr_newa = getExpressionInC(newra._sympy_())
expr_newa

AttributeError: 'Sequence_generic' object has no attribute '_sympy_'

In [263]:
expr_f = getExpressionInC(rf._sympy_())
expr_f[63:]

' (pow(dr, 2)*pow(dt, 2)*(shmem[CI(1,(t),(threadIdx.z),(tidym),(threadIdx.x))] - 2*shmem[CI(1,(t),(threadIdx.z),(threadIdx.y),(threadIdx.x))] + shmem[CI(1,(t),(threadIdx.z),(tidyp),(threadIdx.x))]) - pow(dtheta, 2)*(pow(dr, 2)*(shmem[CI(1,(tm1),(threadIdx.z),(threadIdx.y),(threadIdx.x))] - 2*shmem[CI(1,(t),(threadIdx.z),(threadIdx.y),(threadIdx.x))]) - pow(dt, 2)*(shmem[CI(1,(t),(threadIdx.z),(threadIdx.y),(tidxm))] - 2*shmem[CI(1,(t),(threadIdx.z),(threadIdx.y),(threadIdx.x))] + shmem[CI(1,(t),(threadIdx.z),(threadIdx.y),(tidxp))])))/(pow(dr, 2)*pow(dtheta, 2))'

In [264]:
expr_newf = getExpressionInC(newrf._sympy_())
expr_newf[29:]

AttributeError: 'Sequence_generic' object has no attribute '_sympy_'

In [265]:
expr_g = getExpressionInC(rg._sympy_())
expr_g[29:]

'),(threadIdx.y),(threadIdx.x))] == (pow(dr, 2)*pow(dt, 2)*(shmem[CI(2,(t),(threadIdx.z),(tidym),(threadIdx.x))] - 2*shmem[CI(2,(t),(threadIdx.z),(threadIdx.y),(threadIdx.x))] + shmem[CI(2,(t),(threadIdx.z),(tidyp),(threadIdx.x))]) - pow(dtheta, 2)*(pow(dr, 2)*(shmem[CI(2,(tm1),(threadIdx.z),(threadIdx.y),(threadIdx.x))] - 2*shmem[CI(2,(t),(threadIdx.z),(threadIdx.y),(threadIdx.x))]) - pow(dt, 2)*(shmem[CI(2,(t),(threadIdx.z),(threadIdx.y),(tidxm))] - 2*shmem[CI(2,(t),(threadIdx.z),(threadIdx.y),(threadIdx.x))] + shmem[CI(2,(t),(threadIdx.z),(threadIdx.y),(tidxp))])))/(pow(dr, 2)*pow(dtheta, 2))'

In [266]:
expr_newg = getExpressionInC(newrg._sympy_())
expr_newg[63:]

AttributeError: 'Sequence_generic' object has no attribute '_sympy_'

In [267]:
def writeDefinitionGPU(filename, template, expr, arr):
    import os.path
    data = ""
    #check if header already exist
    if os.path.isfile(filename+".cu"):
        #in that case, just append new declaration
        fin = open(filename+".cu", "rt")
        data = fin.read()
        fin.close()
    else:
        data = "//### THIS IS A GENERATED FILE - DO NOT MODIFY ### \n \
#include \""+filename.split('/')[-1]+".cuh\"\n \
#include \"SharedMem.cuh\"\n"
    string = ''
    if arr == 'a':
        string = '0'
    elif arr == 'F':
        string = '1'
    else:
        string = '2'
    data = data + '\n' + template + """{
    int r = blockIdx.x*blockDim.x + threadIdx.x;
    int theta = blockIdx.y*blockDim.y + threadIdx.y;
    int phi = blockIdx.z*blockDim.z + threadIdx.z;
    int global_phi = phi + phi_offset;
    
    extern __shared__ REAL shmem[];
    if (r<M && theta<N && phi<O){
        fillSharedMemory(shmem, a, F, G, M, N, O, global_phi, r, theta, phi);
    }
    __syncthreads();
    if (r<M && theta<N && phi<O){
        int tidxp = threadIdx.x+1;
        int tidxm = threadIdx.x-1;
        
        int tidyp = threadIdx.y+1;
        int tidym = threadIdx.y-1;        
        
        int tidzp = threadIdx.z+1;
        int tidzm = threadIdx.z-1;
        
        """ + "\n \
\t\tshmem[CI("+string+", tp1, threadIdx.z, threadIdx.y, threadIdx.x)] = " + expr + ";\n\t}\n \
    __syncthreads();\n\
    if (r<M && theta<N && phi<O){\
        copySharedMemoryToGlobal"+str(arr)+"(shmem, a, F, G, M, N, O, global_phi, r, theta, phi, tp1);\n\
    }\n\
}"
    
    fin = open(filename+".cu", "wt")
    fin.write(data)
    fin.close()

def writeDeclarationGPU(filename, template):
    import os.path
    data = ""
    #check if header already exist
    if os.path.isfile(filename+".cuh"):
        #in that case, just append new declaration
        fin = open(filename+".cuh", "rt")
        data = fin.read()
        fin.close()
    else:
        data = "//### THIS IS A GENERATED FILE - DO NOT MODIFY ### \n \
#pragma once\n \
#include <iostream>\n \
#include <unistd.h>\n \
#include <vector>\n \
#include <string>\n \
#include <fstream>\n \
#include <cmath>\n \
#include \"defines.h\"\n \
using namespace std;\n"
    
    data = data + '\n' + template + ";"
    
    fin = open(filename+".cuh", "wt")
    fin.write(data)
    fin.close()

def cleanOldFilesGPU(file):
    import os
    try:
        os.remove(file + ".cu")    
        os.remove(file + ".cuh")
    except:
        return "Already gone."

def writePairs(filename, funcname, expr):
    writeDeclarationGPU(filename, "__global__ void "+funcname+"(REAL *a, REAL *F, REAL *G, size_t l, size_t tp1, size_t t, size_t tm1, size_t tm2, size_t M, size_t N, size_t O, size_t phi_offset, REAL dt, REAL dr, REAL dtheta, REAL dphi, REAL l_1, REAL l_2, REAL lamb, int p, int q, int L)")
    writeDefinitionGPU( filename, "__global__ void "+funcname+"(REAL *a, REAL *F, REAL *G, size_t l, size_t tp1, size_t t, size_t tm1, size_t tm2, size_t M, size_t N, size_t O, size_t phi_offset, REAL dt, REAL dr, REAL dtheta, REAL dphi, REAL l_1, REAL l_2, REAL lamb, int p, int q, int L)", expr, funcname[-1])


def findClosingPar(expr):
    let = 1
    i = 1
    #print(expr)
    while let > 0 and i<len(expr):
        if expr[i] == '(':
            let += 1
        elif expr[i] == ')':
            let -= 1
        i+=1
    return i-1

def replaceAllPow(expr):
    indexes = [m.start() for m in re.finditer('pow', expr)]
    for i in indexes[::-1]:
        open = i+4
        close = findClosingPar(expr[open:])+open
        exponent = expr[open:close].split(',')[-1]
        
        commas = [m.start() for m in re.finditer(',', expr[open:close])]
        #print(expr[open:close])
        argument = expr[open:open+commas[-1]]
        
        rep = ''
        for j in range(int(exponent)):
            rep += argument+'*'
        rep = rep[:-1]
        expr = expr[:i] +'(' + rep + ')'+ expr[close+1:]
    return expr  
        #print(expr)
        #return
        #print(rep)

In [268]:
expr_anp = replaceAllPow(expr_a)
expr_fnp = replaceAllPow(expr_f)
expr_gnp = replaceAllPow(expr_g)

expr_newanp = replaceAllPow(expr_newa)
expr_newfnp = replaceAllPow(expr_newf)
expr_newgnp = replaceAllPow(expr_newg)

In [269]:
cleanOldFilesGPU("../gpu/src/EquationAlfa")
cleanOldFilesGPU("../gpu/src/EquationF")
cleanOldFilesGPU("../gpu/src/EquationG")

writePairs("../gpu/src/EquationAlfa", "computeNexta", expr_anp[63:])
#writePairs("../gpu/src/EquationAlfa", "computeSeconda", expr_newanp[63:])

writePairs("../gpu/src/EquationF", "computeNextF", expr_fnp[63:])
#writePairs("../gpu/src/EquationF", "computeSecondF", expr_newfnp[63:])

writePairs("../gpu/src/EquationG", "computeNextG", expr_gnp[63:])
#writePairs("../gpu/src/EquationG", "computeSecondG", expr_newgnp[63:])